cpp/json_schema_converter.cc

```diff
@@ -1208,6 +1207,176 @@ std::string JSONSchemaConverter::VisitString(
+      std::string pchar = "([\\w\\.~!$&'()*+,;=:@-]|%[0-9A-Fa-f][0-9A-Fa-f])";
+      std::string query_fragment_char = "([\\w\\.~!$&'()*+,;=:@/\\?-]|%[0-9A-Fa-f][0-9A-Fa-f])*";
+      std::string query = "(\\?" + query_fragment_char + ")?";
+      std::string fragment = "(#" + query_fragment_char + ")?";
+      std::string path_abempty = "(/" + pchar + "*)*";
+      std::string path_absolute = "/(" + pchar + "+(/" + pchar + "*)*)?";
+      std::string segment_nz_nc = "([\\w\\.~!$&'()*+,;=@-]|%[0-9A-Fa-f][0-9A-Fa-f])+";
+      std::string path_noscheme = segment_nz_nc + "(/" + pchar + "*)*";
+      std::string userinfo = "([\\w\\.~!$&'()*+,;=:-]|%[0-9A-Fa-f][0-9A-Fa-f])*";
+      std::string host = "([\\w\\.~!$&'()*+,;=-]|%[0-9A-Fa-f][0-9A-Fa-f])*";
+      std::string authority = "(" + userinfo + "@)?" + host + "(:\\d*)?";
+      std::string relative_part =
+          "(//" + authority + path_abempty + "|" + path_absolute + "|" + path_noscheme + ")?";
+      std::string uri_reference_regex_pattern = "^" + relative_part + query + fragment + "$";
+      std::string uri_reference_ebnf = RegexToEBNF(uri_reference_regex_pattern, false);
+      return "\"\\\"\" " + uri_reference_ebnf + " \"\\\"\"";
+    }
+    if (format == "uri-template") {
+      // refer to RFC 6570, section 2
+      std::string literals =
+          "([\\x21\\x23-\\x24\\x26\\x28-\\x3B\\x3D\\x3F-\\x5B\\x5D\\x5F\\x61-\\x7A\\x7E]"
+          "|%[0-9A-Fa-f][0-9A-Fa-f])";
+      std::string op = "[+#\\./;\\?&=,!@|]";
+      std::string varchar = "(\\w|%[0-9A-Fa-f][0-9A-Fa-f])";
+      std::string varname = varchar + "(\\.?" + varchar + ")*";
+      std::string varspec = varname + "(:[1-9]\\d?\\d?\\d?|\\*)?";
+      std::string variable_list = varspec + "(," + varspec + ")*";
+      std::string expression = "\\{(" + op + ")?" + variable_list + "\\}";
+      std::string uri_template_regex_pattern = "^(" + literals + "|" + expression + ")*$";
+      std::string uri_template_ebnf = RegexToEBNF(uri_template_regex_pattern, false);
+      return "\"\\\"\" " + uri_template_ebnf + " \"\\\"\"";
+    }
+    if (format == "json-pointer") {
+      // refer to RFC 6901, section 3
+      std::string json_pointer_regex_pattern =
+          "^(/([\\x00-\\x2E]|[\\x30-\\x7D]|[\\x7F-\\U0010FFFF]|~[01])*)*$";
+      std::string json_pointer_ebnf = RegexToEBNF(json_pointer_regex_pattern, false);
+      return "\"\\\"\" " + json_pointer_ebnf + " \"\\\"\"";
+    }
+    if (format == "relative-json-pointer") {
+      // refer to draft-handrews-relative-json-pointer-01, section 3
+      std::string relative_json_pointer_regex_pattern =
+          "^(0|[1-9][0-9]*)(#|(/([\\x00-\\x2E]|[\\x30-\\x7D]|[\\x7F-\\U0010FFFF]|~[01])*)*)$";
+      std::string relative_json_pointer_ebnf =
+          RegexToEBNF(relative_json_pointer_regex_pattern, false);
+      return "\"\\\"\" " + relative_json_pointer_ebnf + " \"\\\"\"";
+    }
+  }
```

Comment:
You should probably remove the "format" key in `WarnUnsupportedKeywords` below

Reply:
Sure. Actually some sub-types in "format" are still missing, e.g. `iri` and `idn-email`. Maybe we need to notify in detail.